drivers: pwm: nrfx: Add runtime PM and fix case when PWM is not disabled when not used

nordic-krch opened the pull request:
**PWM not disabled when not used**
Shim was not correctly disabling PWM when it was not used. Task STOP was triggered but PWM->ENABLE remained set which caused increased current. Added interrupt and enabled event handler in the nrfx driver to allow disabling PWM on STOPPED event.

**Add runtime PM support**
Rework PM handling to use pm_device_driver_init(). Shim is not using put and get internally as there is no api that disables or stops pwm so it is hard to determine when to put the device. There are cases when PWM peripheral is stopped but PWM is still active because duty cycle is 100% or 0% and pin is driven by GPIO and not PWM.

If user want to use runtime PM with PWM it is possible and getting the device will initialize internal data and putting will suspend by forcing PWM stop if used and setting pins to sleep state. However, from power consumption perspective it is enough to set 0% or 100% duty cycle on all channels.

nordic-krch committed:
drivers: pwm: nrfx: Disable PWM peripheral when not used

Shim was not correctly disabling PWM when it was not used. Task
STOP was triggered but PWM->ENABLE remained set which caused
increased current. Added interrupt and enabled event handler in
the nrfx driver to allow disabling PWM on STOPPED event.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: pwm: nrfx: Improve runtime PM

Rework PM handling to use pm_device_driver_init(). Shim is not using
put and get internally as there is no api that disables or stops
pwm so it is hard to determine when to put the device. There are cases
when PWM peripheral is stopped but PWM is still active because
duty cycle is 100% or 0% and pin is driven by GPIO and not PWM.

If user want to use runtime PM with PWM it is possible and getting
the device will initialize internal data and putting will suspend
by forcing PWM stop if used and setting pins to sleep state. However,
from power consumption perspective it is enough to set 0% or 100%
duty cycle on all channels.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -295,60 +292,68 @@ static int pwm_nrfx_init(const struct device *dev)
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
+		if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
```
This should be `CONFIG_PM_DEVICE`, not `CONFIG_PM_DEVICE_RUNTIME`.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -295,60 +292,68 @@ static int pwm_nrfx_init(const struct device *dev)
-		}
-		ret = pwm_nrfx_init(dev);
+	case PM_DEVICE_ACTION_TURN_ON:
+		ret = pwm_suspend(dev, false);
```
When pinctrl is not called on `PM_DEVICE_ACTION_TURN_ON`, the PWM will end up with unconfigured pins if runtime PM is enabled for it (it is supposed to start in the suspended state then and `PM_DEVICE_ACTION_SUSPEND` is not performed). This can potentially cause some issues as the pins will keep their previous configuration, whatever it was, until the PWM is resumed.

nordic-krch replied:
I reworked it a bit after I've seen work done in TWIM regarding runtime PM. TURN_ON action is removed.

nordic-krch committed:
drivers: pwm: nrfx: Improve device generation macros

Use NRFX_FOREACH_PRESENT macro to iterate over all PWM instances
and create device only for those enabled in the devicetree.
This approach removes need of changing driver code when new
instance id is added.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -387,6 +392,8 @@ static int pwm_nrfx_pm_action(const struct device *dev,
 		IRQ_CONNECT(DT_IRQN(PWM(idx)), DT_IRQ(PWM(idx), priority),    \
 			    nrfx_isr, nrfx_pwm_##idx##_irq_handler, 0);	      \
+		ANOMALY_109_EGU_IRQ_CONNECT(				      \
+				NRFX_PWM_NRF52_ANOMALY_109_EGU_INSTANCE);     \
```
move back.

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -209,27 +219,25 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
+		 * and till that moment, it ignores any start requests,
+		 * so ensure here that it is stopped.
+		 */
+		while (data->stop_requested) {
```
right, change condition for triggering stop from `!data->stop_requested` to  `data->active && !data->stop_requested`. That should handle it.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -284,63 +290,71 @@ static int pwm_nrfx_init(const struct device *dev)
+	struct pwm_nrfx_data *data = dev->data;
 
-	nrfx_pwm_uninit(&config->pwm);
+	if (data->active) {
+		/* PWM is still in use. Force stopping. */
+		data->stop_requested = true;
+		nrfx_pwm_stop(&config->pwm, false);
+		while (data->stop_requested) {
+		}
+	}
```
This could be replaced with:
```
	if (!nrfx_pwm_stopped_check(&config->pwm)) {
		nrfx_pwm_stop(&config->pwm, true);
	}
```
to avoid introducing `active`.

nordic-krch replied:
changed.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -284,63 +290,71 @@ static int pwm_nrfx_init(const struct device *dev)
+		return -EBUSY;
+	}
+
+	ret = pwm_suspend(dev);
```
Why not just call `pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);` here (or better before the nrfx driver is initialized, for consistency with other shims)? That's essentially what is needed, right?

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -209,27 +217,27 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-		data->stop_requested = true;
+		if (data->active && !data->stop_requested) {
+			data->stop_requested = true;
+			nrfx_pwm_stop(&config->pwm, false);
+		}
 	} else {
-		if (data->stop_requested) {
-			data->stop_requested = false;
-
-			/* After a stop is requested, the PWM peripheral stops
-			 * pulse generation at the end of the current period,
-			 * and till that moment, it ignores any start requests,
-			 * so ensure here that it is stopped.
-			 */
-			while (!nrfx_pwm_stopped_check(&config->pwm)) {
-			}
+		/* After a stop is requested, the PWM peripheral stops
+		 * pulse generation at the end of the current period,
+		 * and till that moment, it ignores any start requests,
+		 * so ensure here that it is stopped.
+		 */
+		while (data->stop_requested) {
```
what about the case when we `nrfx_pwm_stop()` is called for the second time on already stopped pwm. There is an assert in nrfx_pwm.c. It can happen if 0/100 duty cycle is set second channel.

nordic-krch replied:
a, ok. We can use `nrfx_pwm_stopped_check(&config->pwm)` there.

anangl replied:
> what about the case when we `nrfx_pwm_stop()` is called for the second time on already stopped pwm. There is an assert in nrfx_pwm.c. It can happen if 0/100 duty cycle is set second channel.

A second stop is not a problem. If you mean `NRFX_ASSERT(p_cb->state != NRFX_DRV_STATE_UNINITIALIZED);`, the condition is still fulfilled then. Actually, the existing code works this way - there's no protection against a second call to `nrfx_pwm_stop()`.

nordic-krch replied:
in that case i simplified change and allow double nrfx_pwm_stop() call.

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -283,64 +282,60 @@ static int pwm_nrfx_init(const struct device *dev)
+		(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
+	}
+
+	memset(dev->data, 0, sizeof(struct pwm_nrfx_data));
```
done

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -283,64 +282,60 @@ static int pwm_nrfx_init(const struct device *dev)
 
 	memset(dev->data, 0, sizeof(struct pwm_nrfx_data));
+	if (IS_ENABLED(CONFIG_PM_DEVICE)) {
+		(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
```
done

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -283,64 +282,53 @@ static int pwm_nrfx_init(const struct device *dev)
 
-	default:
-		return -ENOTSUP;
+	if (IS_ENABLED(CONFIG_PM_DEVICE)) {
```
One last thing, this should rather be `CONFIG_PM_DEVICE_RUNTIME`, because only then the device is expected to be initialized as suspended (when the runtime PM is auto-enabled for it) but the SUSPEND action is not called, right?
